File: node/test/replays.test.ts
import { describe, it, expect } from "vitest";
import request from "supertest";
import { createApp } from "../src/app.js";
import { createTenant, createEndpoint } from "./fixtures.js";

const validBody = { range_start: "2026-01-01T00:00:00.000Z", range_end: "2026-01-02T00:00:00.000Z" };

describe("POST /endpoints/:id/replays", () => {
  it("accepts a replay request and returns 202 with a fresh id and pending_expansion status", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    const app = createApp();

    const res = await request(app)
      .post(`/endpoints/${endpoint.id}/replays`)
      .set("Authorization", `Bearer ${apiKey}`)
      .set("Idempotency-Key", "replay-key-1")
      .send(validBody);

    expect(res.status).toBe(202);
    expect(res.body).toMatchObject({ status: "pending_expansion" });
    expect(typeof res.body.id).toBe("string");
  });

  it("returns the original replay's id when the same Idempotency-Key is reused", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    const app = createApp();

    const first = await request(app)
      .post(`/endpoints/${endpoint.id}/replays`)
      .set("Authorization", `Bearer ${apiKey}`)
      .set("Idempotency-Key", "repeat-replay-key")
      .send(validBody);

    const second = await request(app)
      .post(`/endpoints/${endpoint.id}/replays`)
      .set("Authorization", `Bearer ${apiKey}`)
      .set("Idempotency-Key", "repeat-replay-key")
      .send(validBody);

    expect(second.status).toBe(202);
    expect(second.body.id).toBe(first.body.id);
  });

  it("rejects a replay request with no Idempotency-Key header", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    const app = createApp();

    const res = await request(app)
      .post(`/endpoints/${endpoint.id}/replays`)
      .set("Authorization", `Bearer ${apiKey}`)
      .send(validBody);

    expect(res.status).toBe(400);
  });

  it("rejects a replay request with range_end before range_start", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    const app = createApp();

    const res = await request(app)
      .post(`/endpoints/${endpoint.id}/replays`)
      .set("Authorization", `Bearer ${apiKey}`)
      .set("Idempotency-Key", "reversed-range")
      .send({ range_start: "2026-01-02T00:00:00.000Z", range_end: "2026-01-01T00:00:00.000Z" });

    expect(res.status).toBe(400);
  });

  it("returns 404 for an endpoint that doesn't exist", async () => {
    const { apiKey } = await createTenant();
    const app = createApp();

    const res = await request(app)
      .post(`/endpoints/00000000-0000-0000-0000-000000000000/replays`)
      .set("Authorization", `Bearer ${apiKey}`)
      .set("Idempotency-Key", "missing-endpoint")
      .send(validBody);

    expect(res.status).toBe(404);
  });

  it("returns 404 for an endpoint belonging to a different tenant", async () => {
    const { id: otherTenantId } = await createTenant("other-tenant");
    const foreignEndpoint = await createEndpoint(otherTenantId, ["order.created"]);
    const { apiKey } = await createTenant();
    const app = createApp();

    const res = await request(app)
      .post(`/endpoints/${foreignEndpoint.id}/replays`)
      .set("Authorization", `Bearer ${apiKey}`)
      .set("Idempotency-Key", "cross-tenant")
      .send(validBody);

    expect(res.status).toBe(404);
  });
});

File: go/internal/api/replays_test.go
package api_test

import (
	"net/http"
	"testing"

	"github.com/stretchr/testify/require"

	"webhooks-go/internal/testsupport"
)

var validReplayBody = map[string]any{
	"range_start": "2026-01-01T00:00:00.000Z",
	"range_end":   "2026-01-02T00:00:00.000Z",
}

func TestReplayAcceptsRequestAndReturnsPendingExpansion(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})

	req, err := newJSONRequest(http.MethodPost, ts.URL+"/endpoints/"+endpointID+"/replays", validReplayBody)
	require.NoError(t, err)
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Idempotency-Key", "replay-key-1")

	resp, err := http.DefaultClient.Do(req)
	require.NoError(t, err)
	defer resp.Body.Close()
	require.Equal(t, http.StatusAccepted, resp.StatusCode)

	var body struct {
		ID     string `json:"id"`
		Status string `json:"status"`
	}
	decodeJSON(t, resp, &body)
	require.NotEmpty(t, body.ID)
	require.Equal(t, "pending_expansion", body.Status)
}

func TestReplayIsIdempotentOnRepeatedKey(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})

	postReplay := func() string {
		req, err := newJSONRequest(http.MethodPost, ts.URL+"/endpoints/"+endpointID+"/replays", validReplayBody)
		require.NoError(t, err)
		req.Header.Set("Authorization", "Bearer "+apiKey)
		req.Header.Set("Idempotency-Key", "repeat-replay-key")

		resp, err := http.DefaultClient.Do(req)
		require.NoError(t, err)
		defer resp.Body.Close()
		require.Equal(t, http.StatusAccepted, resp.StatusCode)

		var body struct {
			ID string `json:"id"`
		}
		decodeJSON(t, resp, &body)
		return body.ID
	}

	first := postReplay()
	second := postReplay()
	require.Equal(t, first, second)
}

func TestReplayRejectsMissingIdempotencyKey(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})

	resp := doRequest(t, http.MethodPost, ts.URL+"/endpoints/"+endpointID+"/replays", apiKey, validReplayBody)
	require.Equal(t, http.StatusBadRequest, resp.StatusCode)
}

func TestReplayRejectsRangeEndBeforeRangeStart(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})

	req, err := newJSONRequest(http.MethodPost, ts.URL+"/endpoints/"+endpointID+"/replays", map[string]any{
		"range_start": "2026-01-02T00:00:00.000Z",
		"range_end":   "2026-01-01T00:00:00.000Z",
	})
	require.NoError(t, err)
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Idempotency-Key", "reversed-range")

	resp, err := http.DefaultClient.Do(req)
	require.NoError(t, err)
	defer resp.Body.Close()
	require.Equal(t, http.StatusBadRequest, resp.StatusCode)
}

func TestReplayReturnsNotFoundForMissingEndpoint(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	_, apiKey := testsupport.CreateTenant(t, pool)

	req, err := newJSONRequest(http.MethodPost, ts.URL+"/endpoints/00000000-0000-0000-0000-000000000000/replays", validReplayBody)
	require.NoError(t, err)
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Idempotency-Key", "missing-endpoint")

	resp, err := http.DefaultClient.Do(req)
	require.NoError(t, err)
	defer resp.Body.Close()
	require.Equal(t, http.StatusNotFound, resp.StatusCode)
}

func TestReplayReturnsNotFoundForCrossTenantEndpoint(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	otherTenantID, _ := testsupport.CreateTenant(t, pool)
	foreignEndpointID := testsupport.CreateEndpoint(t, pool, otherTenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	_, apiKey := testsupport.CreateTenant(t, pool)

	req, err := newJSONRequest(http.MethodPost, ts.URL+"/endpoints/"+foreignEndpointID+"/replays", validReplayBody)
	require.NoError(t, err)
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Idempotency-Key", "cross-tenant")

	resp, err := http.DefaultClient.Do(req)
	require.NoError(t, err)
	defer resp.Body.Close()
	require.Equal(t, http.StatusNotFound, resp.StatusCode)
}
